src/main/java/in/org/projecteka/hiu/dataflow/DataFlowService.java

```diff
@@ -2,23 +2,58 @@
 
 import in.org.projecteka.hiu.ClientError;
 import in.org.projecteka.hiu.consent.ConsentRepository;
-import in.org.projecteka.hiu.dataflow.model.DataEntry;
-import in.org.projecteka.hiu.dataflow.model.DataNotificationRequest;
+import in.org.projecteka.hiu.consent.DataFlowRequestPublisher;
+import in.org.projecteka.hiu.dataflow.cryptohelper.CryptoHelper;
 import in.org.projecteka.hiu.dataflow.model.Entry;
+import in.org.projecteka.hiu.dataflow.model.DataNotificationRequest;
+import in.org.projecteka.hiu.dataflow.model.DataFlowRequestKeyMaterial;
+import in.org.projecteka.hiu.dataflow.model.DataEntry;
 import in.org.projecteka.hiu.dataflow.model.Status;
 import lombok.AllArgsConstructor;
+import org.apache.log4j.Logger;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
 
 @AllArgsConstructor
 public class DataFlowService {
     private DataFlowRepository dataFlowRepository;
+    private static final Logger logger = Logger.getLogger(DataFlowRequestPublisher.class);
     private HealthInformationRepository healthInformationRepository;
     private ConsentRepository consentRepository;
+    private CryptoHelper cryptoHelper;
+
+    private Mono<Entry> getDecodedData(DataNotificationRequest dataNotificationRequest, Entry entry){
+        var senderPublicKey = dataNotificationRequest.getKeyMaterial().getDhPublicKey().getKeyValue();
+        var randomKeySender = dataNotificationRequest.getKeyMaterial().getNonce();
+        return dataFlowRepository.getKeys(dataNotificationRequest.getTransactionId())
+                .flatMap(keyPairs -> {
+                    try {
+                        var entryString = getEntryData(senderPublicKey, randomKeySender, entry, keyPairs);
+                        return Mono.just(Entry.builder()
+                                .checksum(entry.getChecksum())
+                                .content(entryString)
+                                .link(entry.getLink())
+                                .media(entry.getMedia())
+                                .build());
+                    } catch (Exception e) {
+                        logger.error("Error while decrypting {exception}", e);
+                        return Mono.error(e);
+                    }
+                });
+    }
+
```

Comment:
Can you create a separate class/static method to handle the decryption which does not need to know where we are getting keys from, get all the values it needs to decrypt any entry given keys?

Reply:
Done